File: src/main/java/gargoyle/ct/util/CTStreamUtil.java
package gargoyle.ct.util;

import gargoyle.ct.log.Log;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.util.Scanner;

@SuppressWarnings("StaticMethodOnlyUsedInOneClass")
public final class CTStreamUtil {
    private static final String DELIMITER = "\\A";

    private CTStreamUtil() {
    }

    @NotNull
    public static String convertStreamToString(@NotNull InputStream is, @NotNull String charsetName) {
        try (Scanner scanner = new Scanner(is, charsetName); Scanner s = scanner.useDelimiter(DELIMITER)) {
            return s.hasNext() ? s.next() : "";
        } catch (RuntimeException ex) {
            Log.error(ex, ex.getMessage());
            throw ex;
        }
    }

    public static void write(@NotNull OutputStream stream, @NotNull String content, @NotNull String charsetName) {
        try (Writer writer = new OutputStreamWriter(stream, charsetName)) {
            writer.write(content);
            writer.flush();
        } catch (UnsupportedEncodingException ex) {
            throw new IllegalArgumentException(ex);
        } catch (IOException ex) {
            throw new ResourceException(ex.getLocalizedMessage(), ex);
        }
    }
}

File: src/main/kotlin/gargoyle/ct/util/CTStreamUtil.kt
package gargoyle.ct.util

import gargoyle.ct.log.Log
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.io.OutputStreamWriter
import java.io.UnsupportedEncodingException
import java.util.Scanner

object CTStreamUtil {
    private const val DELIMITER = "\\A"
    fun convertStreamToString(stream: InputStream, charsetName: String): String {
        try {
            Scanner(stream, charsetName).use { scanner ->
                scanner.useDelimiter(DELIMITER).use { s -> return if (s.hasNext()) s.next() else "" }
            }
        } catch (ex: RuntimeException) {
            Log.error(ex, ex.message ?: "")
            throw ex
        }
    }

    fun write(stream: OutputStream, content: String, charsetName: String) {
        try {
            OutputStreamWriter(stream, charsetName).use { writer ->
                writer.write(content)
                writer.flush()
            }
        } catch (ex: UnsupportedEncodingException) {
            throw IllegalArgumentException(ex)
        } catch (ex: IOException) {
            throw ResourceException(ex.localizedMessage, ex)
        }
    }
}
